docs/design/sql/20-match-jet-job-to-sql.md

```diff
@@ -0,0 +1,189 @@
+of query thread blocking during DML query execution.
+Overall, this way causes a substantial shift for a metrics-related task as a short term goal, necessitating considerable
+effort and time for implementation. But for a long-running perspective it seems reasonable effort.
+
+#### Solution design
+
+##### Client protocol
+
+To enable asynchronous query handling, we propose adding a field to the `SqlExecuteCodec.RequestParameters` to mark 
+requests that should be handled asynchronously. In order to support [Flexible Async Mode](#flexible-async-mode)
+it will be an integer field with the following values:
+- `-1` - synchronous mode (default and for backwards compatibility)
+- `0` - asynchronous mode
+- `> 0` - timeout for flexible async mode before returning async response
+
+Initially only the first 2 values will be supported. Flexible async mode may be implemented in the future if the performance
+tests confirm that latency for short DML queries might be improved that way.
+
+##### Java APIs
+
+In the `SqlService`, we will introduce a new method called `executeAsync`. This method will set the asynchronous flag
+when executing the request, signaling that the query should be handled asynchronously. After executing a query using
+`executeAsync`, the client should check the first arrived result to determine if it produces rows. If rows are produced,
+it will continue with normal execution. If not (user wants to know when the query finishes), then `job.join()` must 
+be explicitly called. For better user experience, job joining may be encapsulated in new `AsyncSqlResult.join()` method.
+
+Note that async execution will not be a default mode for the time being because without flexible async mode
+it may increase latency for short DML queries due to additional roundtrip to the member.
+
+```java
+public interface SqlService {
+    // ...
+
+    /**
+     * Execute the statement in asynchronous way if possible.
+     * @param statement
+     * @return
+     */
+    // also additional convenience variants similar to `execute`
+    // supported only on client-side (SqlClientService).
+    @Nonnull
+    SqlResult executeAsync(@Nonnull SqlStatement statement);
+}
+
+public interface AsyncSqlResult extends SqlResult {
+    /**
+     * @return If the query uses/used Jet job
+     */
+    boolean isJob();
+    
+    /**
+     * Returns job id of job used to execute this query.
+     * <p>
+     * If the job was a light job (interactive query without {@code ANALYZE})
+     * it will be impossible to find the job after the query finished
+     * which requires caution or error handling to avoid race conditions.
+     * 
+     * @throws UnsupportedOperationException if the query did not a use Jet job (isJob() returns false)
+     */
+    long getJobId();
+
+    /**
+     * Wait for query finish.
+     * If the job does not exist it assumes that it already finished.
```

Comment:
should we know if the job was light or not?
should we eagerly join the job (additional request to member) or lazily (the job may no longer exist and we will not be able to get the exception if there was any)?